### ISSUE TITLE
cargo-new ignores $GIT_TEMPLATE_DIR

### ISSUE THREAD

**Issue by BenWiederhake** [rust-lang/cargo#5420]
`cargo new` also initializes git.  Awesome!  Git is configurable through environment variables.  Awesome!  So why does git, when invoked by cargo, suddenly stop listening to its environment?

Steps to reproduce / I tried this:
```
$ mkdir empty_reference_dir
$ export GIT_TEMPLATE_DIR=$(realpath empty_reference_dir)
$ cargo new newproject
```

Expected behavior / I expected to see this happen:
Absence of files/dirs `.git/description`, `.git/hooks/`, and `.git/info/`.
Or more generally: I expected the absence of the files usually found in `/usr/share/git-core/templates`

Actual behavior / Instead, this happened:
It seems like all files from `/usr/share/git-core/templates` were copied.

My best guess: Either cargo crafts its own "clean" environment (why should it?), or cargo's emulation of git is incomplete.

For reference:  A simple `git init` works as I desire, i.e., it obeys `GIT_TEMPLATE_DIR`.  So at least I'm sure I didn't have a brainfart there.

**Comment by alexcrichton** [comment]
I think this is similar to https://github.com/rust-lang/cargo/issues/1295 where it was found that libgit2 doesn't read this env var

**Comment by BenWiederhake** [comment]
You're absolutely right, that's exactly the issue I'm facing here.  I'm sorry I haven't looked closely enough.

So https://github.com/libgit2/libgit2/issues/2910#issuecomment-76943275 has been closed, saying that "In general, libgit2 does not take environment variables into account", which is really weird because it breaks mental models and personal configuration.

I can see only three ways to deal with this:

- Send a bunch of patches to the libgit2 developers which adds that functionality, and promise to maintain it.  They probably won't accept it anyway.  Sigh.
- Change cargo/git-rs/libgit2 so that it obeys "some" git environment flags.  That's even more inconsistent than the current situation.
- Accept that it sucks, and put it into the global git config file (`git config --global …`), and close this issue and #1295 as wont-fix.

Oh well. :/

**Comment by alexcrichton** [comment]
Oh I'd be fine in the meantime to change the default behavior of git2-rs, it makes sense to me at least to follow what `git`-the-CLI does by default

**Comment by ethomson** [comment]
> Send a bunch of patches to the libgit2 developers which adds that functionality, and promise to maintain it. They probably won't accept it anyway. Sigh.

As I mentioned over there, no, we likely will not.  This is not a bug.  Our philosophy is that we are a simple library and there's little way to predict what our users want to do with it.  `GIT_DIR` and `GIT_TEMPLATE_DIR` and the like are meant to instruct your git client.  They are not meant to instruct your [dive log](https://subsurface-divelog.org), for example, or your [package manager](https://docs.julialang.org/en/stable/devdocs/libgit2/).

But I'm not sure why this is a problem; _our_ defaults are only _our_ defaults.  If you want this behavior _in cargo_ then it should be easy to enable.  We would be happy to accept patches that make this experience better and simpler for you, but as mentioned above and in the libgit2 thread - this needs to be explicitly _opt in_.  Please let me know if I can help.  😀

**Comment by dwijnand** [comment]
I think the next actionable step here is to file this issue against git-rs.

**Comment by ethomson** [comment]
> I think the next actionable step here is to file this issue against git-rs.

👍  libgit2 has added some help here - you can now `git_repository_open_with_env` to use these environment variables as hints (to behave more like git itself).  As I mentioned above, it's still opt-in, but I hope it's much easier and less work to get going.

**Comment by dwijnand** [comment]
Opened an issue in git2-rs: https://github.com/alexcrichton/git2-rs/issues/398
